src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs

```diff
@@ -353,15 +353,32 @@ where buildIds.Contains(build.Id)
             return dict.Values.ToList();
         }
 
-        public bool IsProductDependency(int buildId, string repositoryName, string branchName)
+        public bool IsProductDependency(
+            string sourceRepositoryName,
+            string sourceBranchName,
+            string targetRepositoryName,
+            string targetBranchName)
         {
-            return BuildDependencies
-                .Any(d => d.IsProduct
-                    && d.DependentBuildId == buildId
-                    && ((d.Build.GitHubRepository == repositoryName
-                            && d.Build.GitHubBranch == branchName)
-                        || (d.Build.AzureDevOpsRepository == repositoryName
-                            && d.Build.AzureDevOpsBranch == branchName)));
+            var lastBuild = Builds
+                .Where(b =>
+                    (b.GitHubRepository == targetRepositoryName && b.GitHubBranch == targetBranchName)
+                    || (b.AzureDevOpsRepository == targetRepositoryName && b.AzureDevOpsBranch == targetBranchName))
+                .OrderByDescending(b => b.DateProduced)
+                .Select(b => new
+                {
+                    Id = b.Id,
+                    HasProductDependencies = BuildDependencies
+                        .Any(d =>
+                            ((d.DependentBuild.GitHubRepository == sourceRepositoryName
+                                    && d.DependentBuild.GitHubBranch == sourceBranchName)
+                                || (d.DependentBuild.AzureDevOpsRepository == sourceRepositoryName
+                                    && d.DependentBuild.AzureDevOpsBranch == sourceBranchName))
+                            && d.Build == b
+                            && d.IsProduct)
```

Comment:
You are right but I think that this query already does what you've written above. It only checks dependencies for the most recent target repo build (ordered by `DateProduced` column). Result of this whole query is ID of most recent target repo build and `HasProductDependencies` flag saying if there are any build dependencies in this build that come from `source` and have `IsProduct` flag set.

So does it make sense or am I missing something here?

Reply:
Yeah, I think you're right. I think I just was having a hard time reading the code.

Reply:
Can you add a comment here to basically solidify the intention of this query.

Reply:
Yes, I will try to clean it up a bit tomorrow. I think splitting it into two queries for build and dependencies also wouldn't be bad solution because this probably won't have big impact on performance. I'll see what can be done there.

Reply:
I have split the code into several parts and added a few comments. It will still result in single SQL query but hopefully it's a bit clearer now.

Reply:
Thank you! Looks good to me!